kotlinx-coroutines-core/common/src/internal/DispatchedContinuation.kt

```diff
@@ -0,0 +1,287 @@
+        /*
+         * Transitions:
+         * 1) `null` -> claimed, caller will instantiate CC instance
+         * 2) `CC` -> claimed, caller will reuse CC instance
+         * 3) `NON_REUSABLE` -> nothing, proceed, caller will instantiate CC instance
+         */
+        _reusableCancellableContinuation.loop { state ->
+            when (state) {
+                null -> {
+                    if (_reusableCancellableContinuation.compareAndSet(null, REUSABLE_CLAIMED)) {
+                        return null
+                    }
+                }
+                is CancellableContinuationImpl<*> -> {
+                    if (_reusableCancellableContinuation.compareAndSet(state, REUSABLE_CLAIMED)) {
+                        return state as CancellableContinuationImpl<T>
+                    }
+                }
+                NON_REUSABLE -> return null
+                else ->  error("Inconsistent state $state")
+            }
+        }
+    }
+
+    /**
+     * Checks whether there were any attempts to cancel reusable CC while it was in [REUSABLE_CLAIMED] state
+     * and returns [CancelledContinuation] if so, `null` otherwise.
+     * If continuation was cancelled, it becomes non-reusable.
+     *
+     * ```
+     * suspendAtomicCancellableCoroutineReusable { // <- claimed
+     * // Any asynchronous cancellation is "postponed" while this block
+     * // is being executed
+     * } // postponed cancellation is checked here in `getResult`
+     * ```
+     *
+     * See [CancellableContinuationImpl.getResult].
+     */
+    fun checkPostponedCancellation(continuation: CancellableContinuation<*>): CancelledContinuation? {
+        _reusableCancellableContinuation.loop { state ->
+            // not when(state) to avoid Intrinsics.equals call
+            when {
+                state === REUSABLE_CLAIMED -> if (_reusableCancellableContinuation.compareAndSet(REUSABLE_CLAIMED, continuation)) return null
+                state === null -> return null
+                state is CancelledContinuation -> {
+                    require(_reusableCancellableContinuation.compareAndSet(state, NON_REUSABLE))
+                    return state
+                }
+                else -> return null // Is not reusable
+            }
+        }
+    }
+
+    fun makeCancellationNonReusable() {
+        _reusableCancellableContinuation.loop { state ->
+            when (state) {
+                // Do not overwrite cancellation (cancelled CC can't be reused anyway)
+                is CancelledContinuation -> return
+                else -> if (_reusableCancellableContinuation.compareAndSet(state, NON_REUSABLE)) return
+            }
+        }
+    }
+
+    /**
+     * Tries to postpone cancellation if reusable CC is currently in [REUSABLE_CLAIMED] state.
+     * Returns `true` if cancellation is (or previously was) postponed, `false` otherwise.
+     */
+    fun postponeCancellation(cancelled: CancelledContinuation): Boolean {
```

Comment:
There is no need to store the full blown `CancelledContinuation` object in DC state, since it is never fully used, actually. `checkPostponedCancellation` is declared to return `CancelledContinuation`, but only it's `cause` is ever used. Note, that `CancelledContinuation` creation is quite expensive. 

It would be more efficient and clearer to keep here just the corresponding `Throwable` cause. Note, that we need to postpone only cancellation that goes from the parent job and there we always have non-null cause.